chaincode/compute_plan.go

```diff
@@ -0,0 +1,251 @@
+			}
+			traintuple.InModelKeys = append(traintuple.InModelKeys, inModelKey)
+		}
+
+		traintupleKey := traintuple.GetKey()
+
+		// Set the ComputePlanID
+		if i == 0 {
+			traintuple.ComputePlanID = traintupleKey
+			resp.ComputePlanID = traintuple.ComputePlanID
+		} else {
+			traintuple.ComputePlanID = resp.ComputePlanID
+		}
+
+		// Set status: if it has parents it's waiting
+		// if not it's todo and it has to be included in the event
+		if len(computeTraintuple.InModelsIDs) > 0 {
+			traintuple.Status = StatusWaiting
+		} else {
+			traintuple.Status = StatusTodo
+			out := outputTraintuple{}
+			err = out.Fill(db, traintuple, traintupleKey)
+			if err != nil {
+				return resp, err
+			}
+			traintuplesTodo = append(traintuplesTodo, out)
+		}
+
+		err = traintuple.Save(db, traintupleKey)
+		if err != nil {
+			return resp, err
+		}
+		traintupleKeysByID[computeTraintuple.ID] = traintupleKey
+		resp.TraintupleKeys = append(resp.TraintupleKeys, traintupleKey)
+	}
+
+	resp.TesttupleKeys = []string{}
+	for index, computeTesttuple := range inp.Testtuples {
+		traintupleKey, ok := traintupleKeysByID[computeTesttuple.TraintupleID]
+		if !ok {
+			return resp, errors.BadRequest("testtuple index %s: traintuple ID %s not found", index, computeTesttuple.TraintupleID)
+		}
+		testtuple := Testtuple{}
+		testtuple.Model = &Model{TraintupleKey: traintupleKey}
+		testtuple.ObjectiveKey = inp.ObjectiveKey
+		testtuple.AlgoKey = inp.AlgoKey
+
+		inputTesttuple := inputTesttuple{}
+		inputTesttuple.DataManagerKey = computeTesttuple.DataManagerKey
+		inputTesttuple.DataSampleKeys = computeTesttuple.DataSampleKeys
+		inputTesttuple.Tag = computeTesttuple.Tag
+		err = testtuple.SetFromInput(db, inputTesttuple)
+		if err != nil {
+			return resp, err
+		}
+		testtuple.Status = StatusWaiting
+		testtupleKey := testtuple.GetKey()
+		err = testtuple.Save(db, testtupleKey)
+		if err != nil {
+			return resp, err
+		}
+		resp.TesttupleKeys = append(resp.TesttupleKeys, testtupleKey)
+	}
+
+	event := TuplesEvent{}
+	event.SetTraintuples(traintuplesTodo...)
+	err = SendTuplesEvent(db.cc, event)
+	if err != nil {
+		return resp, err
+	}
+
+	return resp, err
+}
+
+func queryComputePlan(db LedgerDB, args []string) (resp outputComputePlanDetails, err error) {
+	inp := inputHash{}
+	err = AssetFromJSON(args, &inp)
+	if err != nil {
+		return
+	}
+	return getComputePlan(db, inp.Key)
+}
+
+func queryComputePlans(db LedgerDB, args []string) (resp []outputComputePlanDetails, err error) {
+	computePlanKeys, err := db.GetIndexKeys("traintuple~computeplanid", []string{"computeplan"})
+	if err != nil {
+		return
+	}
+	for _, key := range computePlanKeys {
+		var computePlan outputComputePlanDetails
+		computePlan, err = getComputePlan(db, key)
+		if err != nil {
+			return
+		}
+		resp = append(resp, computePlan)
+	}
+	return resp, err
+}
+
+func getComputePlan(db LedgerDB, key string) (resp outputComputePlanDetails, err error) {
```

Comment:
Here again `id` and not `key`.

Reply:
Fixed https://github.com/SubstraFoundation/substra-chaincode/pull/22/commits/779e07c9afee0a42f53e86a45dfc2eba434826d2